### ISSUE TITLE
Controlling name of data products

### ISSUE THREAD

**Issue by gijzelaerr** [common-workflow-language/common-workflow-language#668]
Currently, there is no easy way to control the names of file and directory data products. As a workaround, we add steps to our workflows that rename files. This looks ugly to me, and it makes me feel there must be a prettier way to do this. It would be great to have a way to control the name from the workflow level.

here an example of the renaming step:

https://github.com/gijzelaerr/spiel/blob/master/spiel.cwl#L110

related issue:

https://github.com/common-workflow-language/cwltool/issues/696

**Comment by undu** [comment]
I wonder if this could also be a part of a way to move files in folder structure without requiring Javascript. (some people are very reticent to use CWL because of js)

For reference, take this complex step for arranging outputs form multiple steps in a folder: https://github.com/common-workflow-language/common-workflow-language/issues/668

**Comment by mr-c** [comment]
@gijzelaerr In your specific example you can add the random seed to the filename at run time.

**Comment by gijzelaerr** [comment]
whoa a recursive function issue :) (you point to this issue)

**Comment by gijzelaerr** [comment]
@mr-c I could, but that feels also like a hack. And I thought the idea of CWL was to have reusable components, which this CWL file then isn't anymore.

**Comment by undu** [comment]
woops, fixed

**Comment by mr-c** [comment]
@gijzelaerr How would that be non reusable?

**Comment by gijzelaerr** [comment]
For a CWL step where we don't have the random seed as input, but want to use it in the filename. For example the wsclean step.

**Comment by gijzelaerr** [comment]
yes, but other steps don't.

**Comment by gijzelaerr** [comment]
Just to clarify, this is not a request for help with our workflow but a request for improvement/extension for the standard. This is an issue people here at SKA run into, and I am also a bit surprised that CWL has no control over data product naming. Having to build in logic to rename a data product inside a step results in more rigid and less reusable CWL steps.

**Comment by tetron** [comment]
@gijzelaerr I see your "rename" step also copies the file?

You can already logically rename a file by using an expression to create a new file object with the same "location" but a different "basename".  This is what @undu is doing.

**Comment by gijzelaerr** [comment]
investigating...

**Comment by mr-c** [comment]
@gijzelaerr I agree that it would be useful to have a shortcut for this. In the meantime please see https://github.com/gijzelaerr/spiel/pull/8 which accomplishes the same thing without a rename step or ExpressionTool

**Comment by mr-c** [comment]
As an experiment I tried renaming during the `WorkflowStep` with `StepInputExpressionRequirement` and `valueFrom` but then one only has access to the `source`d value (via `self`) and no others so you can only rename to a fixed value:
``` cwl
steps:
  wsclean:
    in:
      ms:
        source: simulator/ms_out
        valueFrom: |
          ${ self.basename = 'new_ms_dirname'; return self; }
```

**Comment by mr-c** [comment]
Biostars user [ionox0](https://www.biostars.org/u/42071/) suggested the following syntax for renaming files more easily:

``` cwl
outputs:
  out_thing:
    glob: *.fastq
    newName: my_new_name.fastq
```
